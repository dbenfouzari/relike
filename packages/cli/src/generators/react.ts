import { paramCase, snakeCase } from "change-case";
import fs from "fs";
import path from "path";

import { replace, replaceFileName } from "../utils/template";

/**
 * Generates component index file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateIndexFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateIndexFile = (key: string, componentName: string) => {
  const templateIndexPath = path.resolve(__dirname, "../templates/index/index.ts.txt");

  const templateIndex = fs.readFileSync(templateIndexPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateIndexPath, key, componentName)),
    content: replace(templateIndex, key, componentName, paramCase),
  };
};

/**
 * Generates component style file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStyleFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStyleFile = (key: string, componentName: string) => {
  const templateStylePath = path.resolve(__dirname, "../templates/react-style/__name__.module.scss.txt");

  const templateStyle = fs.readFileSync(templateStylePath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStylePath, key, componentName)),
    content: replace(templateStyle, key, componentName, snakeCase),
  };
};

/**
 * Generates component story file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStoryFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStoryFile = (key: string, componentName: string) => {
  const templateStoryPath = path.resolve(__dirname, "../templates/react-story/__name__.stories.tsx.txt");

  const templateStory = fs.readFileSync(templateStoryPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStoryPath, key, componentName)),
    content: replace(templateStory, key, componentName),
  };
};

/**
 * Generates component test file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateTestFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateTestFile = (key: string, componentName: string) => {
  const templateTestPath = path.resolve(__dirname, "../templates/react-test/__name__.test.tsx.txt");

  const templateTest = fs.readFileSync(templateTestPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateTestPath, key, componentName)),
    content: replace(templateTest, key, componentName),
  };
};

/**
 * Generates component file
 *
 * @param   {string}  key           The key to replace
 * @param   {string}  componentName The component name
 * @param   {boolean} hasStyle      Does the component have style
 * @example
 * generateComponentFile("name", name, flags.style);
 * @returns {any}                   File path and content.
 */
export const generateComponentFile = (key: string, componentName: string, hasStyle: boolean) => {
  const templateComponentPath = hasStyle
    ? path.resolve(__dirname, "../templates/react-component/__name__(if:style).tsx.txt")
    : path.resolve(__dirname, "../templates/react-component/__name__.tsx.txt");

  const templateComponent = fs.readFileSync(templateComponentPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateComponentPath, key, componentName)),
    content: replace(templateComponent, key, componentName),
  };
};
