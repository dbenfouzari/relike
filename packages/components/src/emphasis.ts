/**
 * Define an emphasis.
 *
 * Used in buttons for example, to know how to handle different buttons.
 */
export enum Emphasis {
  low = "low",
  medium = "medium",
  high = "high",
}
