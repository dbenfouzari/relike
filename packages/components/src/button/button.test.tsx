import { render, screen } from "@testing-library/react";

import Colors from "../colors";
import { Emphasis } from "../emphasis";
import Icons from "../icons";
import Button, { ButtonProps } from "./button";

const setup = (outerProps?: Partial<ButtonProps>) => {
  const props: ButtonProps = {
    children: "Hello Test",
  };

  const mergedProps = { ...props, ...outerProps };

  return render(<Button {...mergedProps} />);
};

describe("Button", () => {
  it("should render successfully with low emphasis", () => {
    setup({ emphasis: Emphasis.low });

    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("should render successfully with medium emphasis", () => {
    setup({ emphasis: Emphasis.medium });

    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("should render successfully with high emphasis", () => {
    setup({ emphasis: Emphasis.high });

    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("should render successfully when disabled", () => {
    setup({ disabled: true });

    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("should render successfully when lightness is low", () => {
    setup({ color: Colors.black });

    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("should render successfully when lightness is high", () => {
    setup({ color: Colors.white, emphasis: Emphasis.high });

    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("should render successfully with icon", () => {
    setup({ icon: Icons.access_time });

    expect(screen.getByRole("button")).toMatchSnapshot();
  });
});
