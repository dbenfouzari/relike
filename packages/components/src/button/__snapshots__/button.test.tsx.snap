// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render successfully when disabled 1`] = `
<button
  class="jsx-1787608142 button button__disabled button__medium"
  disabled=""
>
  <span
    class="elm-0-2-1 elm-d3-0-2-5"
    data-testid="text-elm"
  >
    Hello Test
  </span>
</button>
`;

exports[`Button should render successfully when lightness is high 1`] = `
<button
  class="jsx-1658600500 button button__high"
>
  <span
    class="elm-0-2-1 elm-d5-0-2-7"
    data-testid="text-elm"
  >
    Hello Test
  </span>
</button>
`;

exports[`Button should render successfully when lightness is low 1`] = `
<button
  class="jsx-1850165365 button button__medium"
>
  <span
    class="elm-0-2-1 elm-d4-0-2-6"
    data-testid="text-elm"
  >
    Hello Test
  </span>
</button>
`;

exports[`Button should render successfully with high emphasis 1`] = `
<button
  class="jsx-1787608142 button button__high"
>
  <span
    class="elm-0-2-1 elm-d2-0-2-4"
    data-testid="text-elm"
  >
    Hello Test
  </span>
</button>
`;

exports[`Button should render successfully with icon 1`] = `
<button
  class="jsx-1787608142 button button__medium"
>
  <span
    class="jsx-2439414882 icon material-icons icon"
  >
    access_time
  </span>
  <span
    class="elm-0-2-1 elm-d6-0-2-8"
    data-testid="text-elm"
  >
    Hello Test
  </span>
</button>
`;

exports[`Button should render successfully with low emphasis 1`] = `
<button
  class="jsx-1787608142 button button__low"
>
  <span
    class="elm-0-2-1 elm-d0-0-2-2"
    data-testid="text-elm"
  >
    Hello Test
  </span>
</button>
`;

exports[`Button should render successfully with medium emphasis 1`] = `
<button
  class="jsx-1787608142 button button__medium"
>
  <span
    class="elm-0-2-1 elm-d1-0-2-3"
    data-testid="text-elm"
  >
    Hello Test
  </span>
</button>
`;
